mfrc522: make concurrent-safe

vkorn opened the pull request:
The device itself is not supporting concurrent access. Locking and waiting for a previous operation to finish could introduce a variety of possible dead-locks. While this is not the ideal solution it will prevent hardware-lock for the device if concurrent Read/Write was performed. 
All low-level methods are made private, only 4 left public: 

1. ReadCard
1. ReadAuth
1. WriteCard
1. WriteSectorTrail

I'm not sure whether this is a good approach, tbh, since we're hiding low-level methods which might come in handy sometimes. @jdevelop would be great to have your feedback.

maruel commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -123,25 +123,30 @@ func NewSPI(spiPort spi.Port, resetPin gpio.PinOut, irqPin gpio.PinIn) (*Dev, er
+	spiDev   spi.Conn
+	stop     chan struct{}
+
+	oMu              sync.Mutex
```
Please use the same mutex for both antennaGain and operationTimeout, there's no need to use separate mutex.

maruel commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -176,66 +181,187 @@ func (r *Dev) Halt() error {
+	r.aMu.Lock()
+	defer r.aMu.Unlock()
+
+	if 0 <= gain && gain <= 7 {
```
You can skip the mutex if the value is out of range. Preferably an error would be returned instead of a silent failure. So:
```
func (r *Dev) SetAntennaGain(gain int) error {
  if gain < 0 || gain > 7 {
    return errors.New("mfrc522: invalid gain")
  }
  r.aMu.Lock()
  r.antennaGain = gain
  r.aMu.Unlock()
  return nil
}

```

maruel commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -176,66 +181,188 @@ func (r *Dev) Halt() error {
+
+// ReadCard reads the card sector/block.
+//
+// 	auth - the authentication mode.
```
I don't think this will come up nice in godoc, you can look locally with:
`godoc -http=localhost:6060`
then browse to http://localhost:6060/pkg/periph.io/x/periph/experimental/devices/mfrc522/

maruel commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -176,66 +181,188 @@ func (r *Dev) Halt() error {
 }
 
-// SetAntennaGain configures antenna signal strength.
+// WriteSectorTrail writes the sector trait with sector access bits.
```
trail / trait?

maruel commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -176,66 +181,188 @@ func (r *Dev) Halt() error {
+// 	auth - authentication mode.
+// 	sector - sector to set authentication.
+// 	keyA - the key used for AuthA authentication scheme.
+// 	keyB - the key used for AuthB authentication schemd.
```
scheme

maruel commented:
gohci

vkorn commented:
Fixed typos in comments and added an example_test as suggested.

jdevelop commented:
> All low-level methods are made private, only 4 left public:

I would prefer to at least have an opportunity to use low-level API in the specific cases ( like setting up the antenna gain or using the internal CRC calculator ). I already have some of the dependencies on that, hiding the methods would make it less useful for me.

Can we maybe split the API in 2, the simplified one with locks and "general" methods, and the one with more fine-grained access?

jdevelop commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -174,68 +177,192 @@ func (r *Dev) Halt() error {
+//		MFRC522 SPI Dev private/helper functions
+
+// Checks whether device is already in use.
+func (r *Dev) accessStarted() error {
```
Is it possible to take a function as an argument to and wrap the execution with lock/unlock?

this way it will be more clean to see the boundaries of the invocations, and less error-prone to forget to unlock the mutex.

Something like 
```
func withDeviceAccess(f func() error) error {
  r.xMu.Lock()
  defer r.Xmu.Unlock()
  return f()
}
```

jdevelop replied:
It's more about the style, I would say. Most of the commands that anyone would need to implement are atomic, e.g - a single thread gets access to the device and performs a sequence of the commands on it. So it is better to have a function that wraps that sequence in "transaction" with locking the library, and pass a function that does something useful.

I could be wrong, but the current implementation you have to call
- accessStarted
- do something
- accessStopped

Here we have 3 blocks and it is fairly easy to panic in the code and have accessStopped lock the device forever.

So with the "pass the function and wrap the context" we are guaranteed to have the library lock managed in the single place, and we don't need an auxiliary `bool` flag to represent the device state.

Hope that it makes sense.

maruel replied:
Yes. That said I'm also fine with:

```
r.lockDevice()
defer r.unlockDevice()
```

which essentially boils down to the same thing (unless mistaken?) and I think readers can relate to this as a lock. Basically rename the function and remove the error.

vkorn replied:
The reason I'm not using explicit lock is that Read could be a blocking operation, for example, if timeout set for indefinite wait second thread will be locked until a card is detected in a first thread.

maruel replied:
Ok, let's leave it as that, please still fix the issues in example_test.go and it's fine to go.

jdevelop commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -514,156 +684,19 @@ func (r *Dev) SelectTag(serial []byte) (byte, error) {
-// StopCrypto stops the crypto chip.
-func (r *Dev) StopCrypto() error {
+// stopCrypto stops the crypto chip.
+func (r *Dev) stopCrypto() error {
```
I would move this method (and alike) to the `lowlevel` package rather than making them private.

maruel replied:
That's a fair comment and I'm fine with this idea. I used something similar with lepton, moving all commands to a subpackage cci https://periph.io/x/periph/devices/lepton/cci.

Here I feel this would supersede commands .

jdevelop commented on experimental/devices/mfrc522/mfrc522.go:
```diff
@@ -514,156 +684,19 @@ func (r *Dev) SelectTag(serial []byte) (byte, error) {
-// 	block - the block within sector to authenticate.
-// 	sectorKey - the key to be used for accessing the sector data.
-// 	serial - the serial of the card.
-func (r *Dev) Auth(mode byte, sector, block int, sectorKey [6]byte, serial []byte) (AuthStatus, error) {
```
This function may is required to allow testing the authentication for the sector without reading it, would be nice to keep it,

maruel commented:
Note: I pushed a gofmt fix so you'll have to rebase.

vkorn commented:
As discussed, moved a few most important low-level operations (CRC, auth, r/w) into commands.LowLevel and exposed it under Dev.LowLevel.
Fixed example_test

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package commands
```
What about have the package name imply that it's low level, then drop the "LowLevel" from the package's symbols? Then you can use New, Dev, etc. I'm not particularly fond of 'commands'.

vkorn replied:
What about "raw"? And make raw.Dev to implement Resource (with Halt, as you've mentioned, calling irqPin.Halt).

maruel replied:
I'd prefer "mfrc522raw" or "mfrc522low". This is because when a symbol is referred to in client code, mfrc522raw.Dev makes more sense that raw.Dev, which is not informative.

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+
+// DevRead gets data from a device.
+func (r *LowLevel) DevRead(address int) (byte, error) {
+	data := []byte{((byte(address) << 1) & 0x7E) | 0x80, 0}
```
This reminds me I should fix package mmr to work with full duplex connection. But that's out of scope here.

If you use [...] you create an array instead a slice. But I'm not sure it changes much w.r.t. allocating on stack vs on heap.

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+
+// SetBitmask sets register bit.
+func (r *LowLevel) SetBitmask(address, mask int) error {
+	current, err := r.DevRead(address)
```
Can we cache the bit in some circumstances for performance?

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+}
+
+// LowLevel is a low-level handler of a MFRC522 RFID reader.
+type LowLevel struct {
```
Please implement Halt() and have it call irqPin.Halt(), to unblock WaitForEdge().

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		return nil, err
+	}
+	return []byte{lsb, msb}, nil
```
I think returning a uint16 would be more sensible here. You could rename CRC to CRC16.

maruel commented on experimental/devices/mfrc522/example_test.go:
```diff
@@ -0,0 +1,104 @@
+	defer rfid.Halt()
+
+	// Setting the antenna signal strength.
+	rfid.SetAntennaGain(5)
```
Please check for error everywhere, so that people can copy paste the code safely and this encourages good code.

maruel commented on experimental/devices/mfrc522/example_test.go:
```diff
@@ -0,0 +1,104 @@
+			// Some devices tend to send wrong data while RFID chip is already detected
+			// but still "too far" from a receiver.
+			// Especially some cheap CN clones which you can find on GearBest, AliExpress, etc.
+			// This will suppress such errors.
```
What about having a clear distinction in the error returned by ReadCard() to distinguish between bus error and logical error? Is it possible to distinguish?

vkorn replied:
Unfortunately, you can get a whole variety of errors. I have devices from 3 different manufacturers and those two which are from China are sending different signals every time when a card is on the edge of antenna radius :(

maruel replied:
I specifically mean that I would expect in this case the error to be one of:
- AuthReadFailure / "can not authenticate"
- In LowLevel.Auth() differentiate when `n&0x08 == 0` triggers vs DevRead() returns an error.
- "expected 16 bytes, actual %d"
- "can't read data after 2000 loops"
- "IRQ error"

vs sysfs SPI driver reporting an error.

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+// 	sectorKey - the key to be used for accessing the sector data.
+// 	serial - the serial of the card.
+func (r *LowLevel) Auth(mode byte, blockAddress byte, sectorKey [6]byte, serial []byte) (AuthStatus, error) {
+	buffer := make([]byte, 2)
```
Here's a way to not allocate memory on the heap, but that's really a micro optimization so you can ignore. :)
```
var b [32]byte{mode, blockAddress}
buf := append(b[:2], sectorKey[:]...)
buf = append(buf, serial...)
```

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+// 	command - the command register
+// 	data - the data to write out to the card using the authenticated sector.
+func (r *LowLevel) CardWrite(command byte, data []byte) ([]byte, int, error) {
+	var backData []byte
```
You can move these two at line 252.

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+	i := 2000
+	n := byte(0)
+
+	for ; i > 0; i-- {
```
What about a time based check? This code is literally dependent on the I²C bus speed...

maruel commented on experimental/devices/mfrc522/commands/low_level.go:
```diff
@@ -0,0 +1,293 @@
+		}
+
+		if n > 16 {
+			n = 16
```
Can you document why cap at 16 bytes?

maruel commented:
gohci
